### PR TITLE
Add bzlmod support + tests

### PR DESCRIPTION
**Pull request by zaucy**
Updated and refactored bazel support in preparation to submit to the bazel central registry https://registry.bazel.build/

Additionally I've added:
 * bazel tests with GitHub actions
 * GitHub action to automatically create a release archive required for bzlmod. Modules get rejected without this. See reason here: https://blog.bazel.build/2023/02/15/github-archive-checksum.html

### CODE REVIEW

File: test/BUILD.bazel
Comment (zaucy) [inline test/BUILD.bazel:10]: I didn't add all the tests, because I didn't want it to be a burden to update this. Since EnTT is already well covered test-wise with CMake I didn't see it as necessary. Especially since the bazel BUILD files previously did not even have tests.

File: BUILD.bazel
Comment (zaucy) [inline BUILD.bazel:7]: I moved this target to the `src` directory to avoid using `strip_include_prefix`. There are issues with the virtual file paths when generating compile_commands.json for editor integration so I like to avoid it if possible.

File: README.md
Comment (zaucy) [inline README.md:332]: This link won't exist until published to the BCR (https://github.com/bazelbuild/bazel-central-registry)

File: test/BUILD.bazel
Comment (skypjack) [reply test/BUILD.bazel:10]: Maybe meta rather than container, since it's probably the second most used submodule? Also signal looks like a good candidate.

File: BUILD.bazel
Comment (skypjack) [reply BUILD.bazel:7]: I barely get what you mean 😅 I guess I'll trust you here. 👍

File: .github/workflows/bazel.yml
Comment (skypjack) [inline .github/workflows/bazel.yml:5]: Same as above, see the `fail_ci_if_error: false` comment.

File: README.md
Comment (skypjack) [inline README.md:330]: _Bazels_ 👈 is there an extra _s_?

File: README.md
Comment (skypjack) [inline README.md:335]: Should we remove the reference in the _Requirements_ section or is it still valid?

Comment (zaucy) [reply README.md:335]: I think it's still valid!

File: .github/workflows/bazel-release-archive.yml
Comment (zaucy) [reply .github/workflows/bazel-release-archive.yml:7]: I couldn't find documentation for `fail_ci_if_error` did you mean
```yaml
continue-on-error: true
```
Or maybe I'm blind 😅

File: README.md
Comment (zaucy) [reply README.md]: should've been `'s` thanks!

File: test/BUILD.bazel
Comment (zaucy) [reply test/BUILD.bazel]: the meta tests are failing 😕 
https://github.com/skypjack/entt/actions/runs/6027899063/job/16354072478?pr=1057

I browsered through the cmake files to see if there's any special flags that I'm not passing to bazel, but I didn't see anything.

Is there something obvious I'm missing that you can think of?

File: .github/workflows/bazel-release-archive.yml
Comment (skypjack) [reply .github/workflows/bazel-release-archive.yml:7]: Oh, yeah, sorry, `fail_ci_if_error` is something like the codecov counterpart for the same feature. My fault. 🙏

File: test/BUILD.bazel
Comment (skypjack) [reply test/BUILD.bazel]: No idea actually. Are you compiling tests as shared library? In this case, the context could be different and the tests would fail.

Comment (zaucy) [reply test/BUILD.bazel]: No I'm not, but that gives me some clues. I'll look into it soon.

Comment (zaucy) [reply test/BUILD.bazel]: I finally got around to looking at this. I didn't understand that the meta tests needed to be compiled separately like the CMake config was doing. That fixed it!

Comment (skypjack) [reply test/BUILD.bazel]: Oh, did you compile all tests in a single executable before? If this is the case, I strongly suggest to split them as I do in the cmake file then.
I want to _clean up_ things and make it possible to compile everything at once but I never found the time to work on this task yet. I'm sorry for the inconvenience.
Let me know if it's ok to merge or if we need to split it a little further. Thanks. 👍

Comment (zaucy) [reply test/BUILD.bazel]: yes they're all separate now! Everything is good to go.

**Review by skypjack — APPROVED** [review]
I'm ok with these changes. However, let me check a thing with you before merging.
I see a file within the `src` dir as well as many other all around the codebase, such as those within `test'.
Is there any way in bazel to _centralize_ things and avoid this kind of _pollution_?
Not a problem on itself but I dislike the way cmake works already because it has the same _problem_ and I prefer to keep dirs clean as much as possible. So, if there exists a way, I'd love to explore it, otherwise nvm and we can merge it as-is.
Thanks and sorry for bothering with my freaking obsessions. 😅

**Comment by zaucy** [comment]
I understand, polluting the repo with files that you're not familiar with can be annoying :P

I added `src/BUILD.bazel` to avoid using the `strip_include_prefix` feature since it has a slight negative impact for developer experience. I can remove the file and use `strip_include_prefix` if it's annoying.

I can also consolidate the `BUILD.bazel` tests into `test/BUILD.bazel` instead if you'd like. It would make all the targets at the root of the test module which is fine, but not preferred for bazel users. If you want I can make that change.

Unfortunately with bazel I couldn't, for instance, put everything in 1 folder separate from everything. Bazel's build files really want to be where the sources are.

**Comment by zaucy** [comment]
the current bazel test targets are this:

```
//entt/common:common
//entt/entity:component
//entt/entity:entity
//entt/entity:group
//entt/entity:handle
//entt/entity:helper
//entt/entity:observer
//entt/entity:organizer
//entt/entity:registry
//entt/entity:runtime_view
//entt/entity:sigh_mixin
//entt/entity:snapshot
//entt/entity:sparse_set
//entt/entity:storage
//entt/entity:storage_entity
//entt/entity:view
//entt/meta:meta_any
//entt/meta:meta_base
//entt/meta:meta_container
//entt/meta:meta_context
//entt/meta:meta_conv
//entt/meta:meta_ctor
//entt/meta:meta_data
//entt/meta:meta_dtor
//entt/meta:meta_func
//entt/meta:meta_handle
//entt/meta:meta_pointer
//entt/meta:meta_prop
//entt/meta:meta_range
//entt/meta:meta_template
//entt/meta:meta_type
//entt/meta:meta_utility
```

I could make them flat like I mentioned and add some prefix instead of using the folder so it's something like this:

```

//:entt_common_common
//:entt_entity_component
//:entt_entity_entity
//:entt_entity_group
//:entt_entity_handle
//:entt_entity_helper
//:entt_entity_observer
//:entt_entity_organizer
//:entt_entity_registry
//:entt_entity_runtime_view
//:entt_entity_sigh_mixin
//:entt_entity_snapshot
//:entt_entity_sparse_set
//:entt_entity_storage
//:entt_entity_storage_entity
//:entt_entity_view
//:entt_meta_meta_any
//:entt_meta_meta_base
//:entt_meta_meta_container
//:entt_meta_meta_context
//:entt_meta_meta_conv
//:entt_meta_meta_ctor
//:entt_meta_meta_data
//:entt_meta_meta_dtor
//:entt_meta_meta_func
//:entt_meta_meta_handle
//:entt_meta_meta_pointer
//:entt_meta_meta_prop
//:entt_meta_meta_range
//:entt_meta_meta_template
//:entt_meta_meta_type
//:entt_meta_meta_utility
```

Not my preference, but I can do that!

**Comment by skypjack** [comment]
Nah, not a problem then.
I guess I'll keep it intact as long as you're around, given that I know almost nothing about bazel. 😅 
Then 🤷‍♂️ let's see, without any help from the community I'll probably drop the support, dunno.

**Comment by zaucy** [comment]
I'll be around! We're using EnTT with bazel so it's a must for us :) I'm happy to maintain. I usually watch the repo for changes, but feel free to @zaucy if something comes up.

**Comment by skypjack** [comment]
@zaucy before I spend my day looking into the failure, do you happen to know already what the cause is for [this error](https://github.com/skypjack/entt/actions/runs/7189044405/job/19590505377?pr=1097)? 🙂

**Comment by zaucy** [comment]
@skypjack should be fixed here https://github.com/skypjack/entt/pull/1098

**Pull request merged**
